### PR TITLE
[Recognizers-Text] improve the validation in the dateTime model

### PR DESCRIPTION
**Pull request by FedericoBernal**

### CODE REVIEW

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java [hunk 1]
```diff
@@ -35,55 +34,54 @@ public DateTimeTest(TestCase currentCase) {
...
-                    (List<Map<String, Object>>)expected.resolution.get(ResolutionKey.ValueSet),
-                    (List<Map<String, Object>>)actual.resolution.get(ResolutionKey.ValueSet));
+    @Override
+    protected void assertResolutionKeys(ModelResult expected, ModelResult actual, TestCase currentCase, List<String> testResolutionKeys) {
+        if (expected.resolution.get(ResolutionKey.ValueSet) instanceof List) {
+            List<HashMap<String, String>> expectedValueSet = (List<HashMap<String, String>>) expected.resolution.get(ResolutionKey.ValueSet);
+            List<HashMap<String, String>> actualValueSet = (List<HashMap<String, String>>) actual.resolution.get(ResolutionKey.ValueSet);
+
+            IntStream.range(0, expectedValueSet.size())
+                .forEach(idx -> {
```
Comment (VictorGrycuk) [inline Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java:64]: I would extract the code inside the foreach to a function to make it more readable

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ResolutionKey.java [hunk 2]
```diff
@@ -9,4 +11,9 @@
     public static final String IsoCurrency = "isoCurrency";
     public static final String Offset = "offset";
     public static final String RelativeTo = "relativeTo";
+    public static final String Timex = "timex";
+    public static final String Start = "start";
+    public static final String End = "end";
+    public static final String Mod = "mod";
+    public static final String SourceEntity = "sourceEntity";
```
Comment (Batta32) [inline Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ResolutionKey.java]: These values should be removed, these are not related to the `ResolutionKey`.

These are related to the DateTimeResolutionKey and they are already added in [DateTimeResolutionKey](https://github.com/microsoft/Recognizers-Text/blob/master/Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/DateTimeResolutionKey.java) following [C#](https://github.com/microsoft/Recognizers-Text/blob/08b488ae1c83127eb93b26d381b8d668b548a620/.NET/Microsoft.Recognizers.Text.DateTime/DateTimeResolutionKey.cs)

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ResolutionKey.java [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package com.microsoft.recognizers.text;
 
+import java.security.PublicKey;
+
```
Comment (Batta32) [inline Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ResolutionKey.java]: Are we using this?

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java [hunk 4]
```diff
@@ -35,55 +34,54 @@ public DateTimeTest(TestCase currentCase) {
...
 
-            Assert.assertNotNull(getMessage(currentCase,
-                    ResolutionKey.ValueSet), actual.resolution.get(ResolutionKey.ValueSet));
+    private List<String> getKeysToTest(TestCase currentCase) {
+        switch (currentCase.modelName) {
+            case "DateTimeModelExtendedTypes":
+            case "DateTimeModelSplitDateAndTime":
+                return Arrays.asList(ResolutionKey.Timex, ResolutionKey.Type, ResolutionKey.Value, ResolutionKey.Start, ResolutionKey.End, ResolutionKey.Mod);
+            default:
+                return Arrays.asList(ResolutionKey.Timex, ResolutionKey.Type, ResolutionKey.Value, ResolutionKey.Start, ResolutionKey.End, ResolutionKey.Mod, ResolutionKey.SourceEntity);
```
Comment (Batta32) [inline Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java]: As we will remove most of them from `ResolutionKey`, these should be changed to consume the `DateTimeResolutionKey` consts

**Review by Batta32 — CHANGES_REQUESTED** [review]

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java (hunk 1 above)
Comment (Batta32) [reply Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java:64]: Agree, but it's a nit. At the meantime, we can add some comments to explain the functionality.

**Review by Batta32 — COMMENTED** [review]
At this level this is approved, as soon as we have the tests passing correctly, we can merge the PR

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseMergedDateTimeParser.java [hunk 5]
```diff
@@ -627,8 +627,7 @@ private void resolveAmPm(Map<String, Object> resolutionDic, String keyName) {
             String timex = (String)resolutionDic.get(DateTimeResolutionKey.Timex);
             timex = timex != null ? timex : "";
 
-            resolutionDic.remove(keyName);
-            resolutionDic.put(keyName + "Am", resolution);
+            resolutionDic.replace(keyName + "Am", resolution);
```
Comment (Batta32) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseMergedDateTimeParser.java]: Not the correct way to replace the keyname, the key cannot be replaced

**Review by Batta32 — CHANGES_REQUESTED** [review]

**Review by Batta32 — APPROVED** [review]

**Pull request merged**